R/compare.R

```diff
@@ -101,12 +146,54 @@ compare <- function(x, y, ...,
     ignore_attr = ignore_attr,
     ignore_encoding = ignore_encoding,
     ignore_formula_env = ignore_formula_env,
-    ignore_function_env = ignore_function_env
+    ignore_function_env = ignore_function_env,
+    ignore_name_order = ignore_name_order,
+    ignore_NULLs = ignore_NULLs,
+    ignore_private = ignore_private
   )
+
+  opts$user_specified <- c(intersect(names(match.call()),
+                                   names(opts)),
+                           "user_specified")
+
   out <- compare_structure(x, y, paths = c(x_arg, y_arg), opts = opts)
   new_compare(out, max_diffs)
 }
 
+sort_named_parts <- function(x) {
+  nx <- names(x)
+  if (!is.null(nx)) {
+    nonempty <- nchar(nx) > 0
+    ox <- order(nx[nonempty])
+    x[nonempty] <- x[nonempty][ox]
+    names(x)[nonempty] <- nx[nonempty][ox]
+  }
+  x
+}
+
+zap_nulls <- function(x) {
+  if (is_list(x) || is_pairlist(x)) {
+    for (i in rev(seq_along(x)))
+      if (is.null(x[[i]]))
+        x[[i]] <- NULL
+  }
+  x
+}
+
+zap_private <- function(x, private, path) {
```

Comment:
Can you explain the logic here? I would've expected it to just be something like this:

```R
x[!grepl(private, names2(x)]
```

Reply:
I've simplified this function a bit using your suggestions above, but it's still somewhat complicated.  

My thinking is that your object might be a list containing other objects, and you might want to specify that certain parts of those component objects are private, with different rules for each component.  So you might want to ignore 
`x$a$private`, but not `x$b$private` or `x$a$c$private`.  You could set `waldo_opts` on `x$a` and `x$a$c`, but it's simpler to just set it on `x`.